File: src/modules/controllers/SeatController.js
class SeatController {
    constructor(seatModel) {
        this.seatModel = seatModel;
        this.LOCK_DURATION = 5 * 60 * 1000;
    }

    isLockExpired(seat) {
        if (!seat.lockTimestamp) return true;
        return (Date.now() - seat.lockTimestamp) > this.LOCK_DURATION;
    }

    selectSeat(userId, seatNumber) {
        const seat = this.seatModel.getSeat(seatNumber);

        if (!seat) return { success: false, message: "Invalid seat." };

        if (seat.status === 'booked') {
            return { success: false, message: "Seat already booked." };
        }

        if (seat.status === 'locked' && !this.isLockExpired(seat)) {
            return { success: false, message: "Seat is locked." };
        }

        this.lockSeat(userId, seatNumber);
        return { success: true, message: `Seat ${seatNumber} locked.` };
    }

    lockSeat(userId, seatNumber) {
        this.seatModel.updateSeat(seatNumber, {
            status: 'locked',
            lockedBy: userId,
            lockTimestamp: Date.now()
        });
    }

    releaseSeat(seatNumber) {
        this.seatModel.updateSeat(seatNumber, {
            status: 'available',
            lockedBy: null,
            lockTimestamp: null
        });
    }

    bookSeat(seatNumber) {
        this.seatModel.updateSeat(seatNumber, {
            status: 'booked',
            lockedBy: null,
            lockTimestamp: null
        });
    }
}

module.exports = SeatController;

File: src/modules/Booking&Seat Module/controller/SeatController.java

package controller;

import model.Seat;
import model.SeatModel;

public class SeatController {

    private SeatModel seatModel;
    private final long LOCK_DURATION = 5 * 60 * 1000;

    public SeatController(SeatModel seatModel){
        this.seatModel = seatModel;
    }

    private boolean isExpired(Seat seat){
        if(seat.lockTimestamp == 0) return true;
        return System.currentTimeMillis() - seat.lockTimestamp > LOCK_DURATION;
    }

    public String selectSeat(String userId, String seatNumber){
        Seat s = seatModel.getSeat(seatNumber);

        if(s == null) return "Invalid seat";

        if(s.status.equals("booked")) return "Already booked";

        if(s.status.equals("locked") && !isExpired(s))
            return "Seat locked";

        s.status = "locked";
        s.lockedBy = userId;
        s.lockTimestamp = System.currentTimeMillis();

        return "Seat locked";
    }

    public void bookSeat(String seatNumber){
        Seat s = seatModel.getSeat(seatNumber);
        s.status = "booked";
        s.lockedBy = null;
        s.lockTimestamp = 0;
    }
}
